File: xarm_moveit_demo/scripts/moveit_planning_scene_demo.py
#!/usr/bin/env python
# -*- coding: utf-8 -*-
import rospy, sys, os
import moveit_commander
from moveit_commander import MoveGroupCommander, PlanningSceneInterface
from moveit_msgs.msg import CollisionObject, AttachedCollisionObject
from shape_msgs.msg import SolidPrimitive
from geometry_msgs.msg import PoseStamped, Pose
from math import pi
from copy import deepcopy
import tf
from tf.transformations import *

class MoveItObstaclesDemo:
    def __init__(self):

        # 初始化Python API 依赖的moveit_commanderC++系统，需放在前面
        moveit_commander.roscpp_initialize(sys.argv)

        # 初始化ROS节点，节点名为'moveit_obstacles_demo'
        rospy.init_node('moveit_planning_scene_demo', anonymous=True)

        # 初始化场景对象
        scene = PlanningSceneInterface()
        rospy.sleep(1)

        # 初始化需要控制的规划组。手臂xarm，手爪gripper
        arm = moveit_commander.MoveGroupCommander('xarm')
        gripper = moveit_commander.MoveGroupCommander('gripper')

        # 设置允许的最大速度
        arm.set_max_velocity_scaling_factor(0.8)
        gripper.set_max_velocity_scaling_factor(0.8)

        # 当运动规划失败后，允许重新规划
        arm.allow_replanning(True)

        # 设置目标位姿为Home
        arm.set_named_target('Home')
        arm.go()

        print "============ Press `Enter` to add objects to the planning scene ..."
        raw_input()
        # 将桌子(长方体桌面)添加到规划场景中
        table_id = 'table'
        table_size = [1.0, 1.2, 0.01]
        table_pose = PoseStamped()
        table_pose.header.frame_id = 'base_link'
        table_pose.pose.position.x = 0.0
        table_pose.pose.position.y = 0.0
        table_pose.pose.position.z =  -table_size[2]/2
        table_pose.pose.orientation.w = 1.0
        scene.add_box(table_id, table_pose, table_size)
        # 判断是否已经将table添加到规划场景中
        if self.wait_for_state_update(table_id,scene,obstacle_is_known=True):
            rospy.loginfo("The Table has been successfully added.")
        else:
            rospy.loginfo("Failed to add the Table.")

        # 将一个长方体box添加到规划场景中
        # 设置长方体box1的ID、尺寸和位姿
        box_id = 'box'
        box_size = [0.25, 0.3, 0.04]
        box_pose = PoseStamped()
        box_pose.header.frame_id = 'base_link'
        box_pose.pose.position.x = 0.4
        box_pose.pose.position.y = 0
        box_pose.pose.position.z = box_size[2]/2 + 0.37
        box_pose.pose.orientation.w = 1.0
        # 调用add_box()函数将box添加到规划场景中
        scene.add_box(box_id, box_pose, box_size)

        # 判断是否已经将box添加到规划场景中
        if self.wait_for_state_update(box_id,scene,obstacle_is_known=True):
            rospy.loginfo("The Box has been successfully added.")
        else:
            rospy.loginfo("Failed to add the Box")

        # 将一个圆柱体添加到规划场景中
        cylinder_object = CollisionObject()
        cylinder_object.header.frame_id = "base_link"
        cylinder_object.id = "cylinder"
        cylinder_primitive = SolidPrimitive()
        cylinder_primitive.type = cylinder_primitive.CYLINDER
        cylinder_primitive.dimensions = [0.12, 0.015]
        cylinder_pose = Pose()
        cylinder_pose.position.x = 0.4
        cylinder_pose.position.y = 0
        cylinder_pose.position.z = cylinder_primitive.dimensions[0]/2.0 + box_size[2]/2 + 0.37
        cylinder_pose.orientation.w = 1.0
        cylinder_object.primitives =[cylinder_primitive]
        cylinder_object.primitive_poses =[cylinder_pose]
        cylinder_object.operation = cylinder_object.ADD
        scene.add_object(cylinder_object)
        if self.wait_for_state_update(cylinder_object.id,scene,obstacle_is_known=True):
            rospy.loginfo("The Cylinder has been successfully added.")
        else:
            rospy.loginfo("Failed to add the Cylinder.")

        print " "
        print "============ Press `Enter` to open gripper, move the arm and close gripper..."
        raw_input()
        # 手爪张开
        joint_positions = [0.65, 0.65]
        gripper.set_joint_value_target(joint_positions)
        gripper.go()
        rospy.sleep(1)
        # 设置机械臂末端执行器的目标位姿，参考坐标系为base_link
        target_pose = PoseStamped()
        target_pose.header.frame_id = "base_link"
        target_pose.header.stamp = rospy.Time.now()
        target_pose.pose.position.x = 0.41
        target_pose.pose.position.y = 0
        target_pose.pose.position.z = cylinder_primitive.dimensions[0]/2.0 + box_size[2]/2 + 0.37
        target_pose.pose.orientation.w = 1
        arm.set_start_state_to_current_state()
        # 设置目标target_pose并让机械臂运动到此目标
        arm.set_pose_target(target_pose)
        arm.go()
        rospy.sleep(1)
        # 闭合手爪
        joint_positions = [0.2, 0.2]
        gripper.set_joint_value_target(joint_positions)
        gripper.go()
        rospy.sleep(1)

        print " "
        print "============ Press `Enter` to attach the Cylinder to the XBot-Arm robot ..."
        raw_input()
        # 使用attach_object()函数把cylinder附着到机械臂末端执行器上
        attach_object = AttachedCollisionObject()
        attach_object.link_name = "gripper_centor_link"
        attach_object.object = cylinder_object
        scene.attach_object(attach_object)
        # 判断cylinder是否附着成功
        if self.wait_for_state_update(cylinder_object.id,scene,object_is_attached=True):
            rospy.loginfo("The cylinder has been successfully attached.")
        else:
            rospy.loginfo("Failed to attach the cylinder.")

        print " "
        print "============ Press `Enter` to move the arm to a position 25 cm down  ..."
        raw_input()
        arm.shift_pose_target(2,-0.25, "gripper_centor_link")
        arm.go()
        rospy.sleep(1)

        print " "
        print "============ Press `Enter` to detach the Cylinder from the XBot-Arm robot ..."
        raw_input()

        # 使用remove_attached_object()函数将cylinde从机械臂上分离
        scene.remove_attached_object("gripper_centor_link", cylinder_object.id)

        rospy.sleep(1)

        print " "
        print "============ Press `Enter` to remove the obstacles ..."
        raw_input()
        # 从规划场景里移除桌面、长方体和圆柱体
        scene.remove_world_object(table_id)
        scene.remove_world_object(box_id)
        scene.remove_world_object(cylinder_object.id)
        rospy.sleep(1)

        print " "
        print "============ Press `Enter` to move to Home and exit the program ..."
        raw_input()
        arm.set_named_target('Home')
        arm.go()
        rospy.sleep(1)
        joint_positions = [0.0, 0.0]
        gripper.set_joint_value_target(joint_positions)
        gripper.go()
        rospy.sleep(1)

        # 干净地关闭moveit_commander并退出程序
        moveit_commander.roscpp_shutdown()
        moveit_commander.os._exit(0)

    # 判断是否已成功添加物体到规划场景，或是否成功把物体附着到机械臂上
    def wait_for_state_update(self,obstacle_name,scene,obstacle_is_known=False, object_is_attached=False, timeout=4):
        start = rospy.get_time()
        seconds = rospy.get_time()
        # 在4s的时间内，循环判断我们想要的规划场景状态是否更新成功，若成功，返回True
        while (seconds - start < timeout) and not rospy.is_shutdown():
          # 判断物体是否attach成功
          attached_objects = scene.get_attached_objects([obstacle_name])
          is_attached = len(attached_objects.keys()) > 0
          # 判断传入的物体是否已经在规划场景中，即物体是否添加成功
          is_known = obstacle_name in scene.get_known_object_names()
          # 判断是否是我们想要的规划场景更新状态
          if (object_is_attached == is_attached) and (obstacle_is_known == is_known):
            return True
          # sleep0.1s，为物体添加或附着留出时间。
          rospy.sleep(0.1)
          seconds = rospy.get_time()
        return False

if __name__ == "__main__":
    try:
        MoveItObstaclesDemo()
    except rospy.ROSInterruptException:
        pass

File: xarm_moveit_demo/src/moveit_planning_scene_demo.cpp
#include <iostream>
#include <ros/ros.h>
#include <moveit/move_group_interface/move_group_interface.h>
#include <moveit/planning_scene_interface/planning_scene_interface.h>
#include <moveit_msgs/AttachedCollisionObject.h>
#include <moveit_msgs/CollisionObject.h>
// 判断是否已成功添加物体到规划场景，或是否成功把物体附着到机械臂上
bool waitForStateUpdate(std::string obstacle_name, moveit::planning_interface::PlanningSceneInterface& scene, bool obstacle_is_known=false, bool object_is_attached=false, double  timeout=4){
  ros::Time start_time = ros::Time::now();
  ros::Time seconds = ros::Time::now();
  bool is_attached, is_known;
  while(seconds - start_time <ros::Duration(timeout) && !ros::isShuttingDown()){
    std::vector<std::string> attached_object_ids;
    attached_object_ids.push_back(obstacle_name);
    // 获取由指定ID列表标记的附着对象物体
    auto attached_objects = scene.getAttachedObjects(attached_object_ids);
    is_attached = attached_objects.size();
    // 获取规划场景中已有的所有物体的ID
    std::vector<std::string> world_object_names = scene.getKnownObjectNames();
    if(std::find(world_object_names.begin(),world_object_names.end(), obstacle_name) != world_object_names.end()){
      is_known =true;
    }
    else {
      is_known =false;
    }
    if(object_is_attached == is_attached && obstacle_is_known == is_known)
      return true;
    usleep(100000);
    seconds = ros::Time::now();
  }
  return false;
}


int main(int argc, char **argv){
  ros::init(argc, argv, "moveit_planning_scene_demo");
  ros::NodeHandle nh;
  ros::AsyncSpinner spinner(1);
  spinner.start();
  moveit::planning_interface::MoveGroupInterface arm("xarm");
  moveit::planning_interface::MoveGroupInterface gripper("gripper");
  // 创建PlanningSceneInterface对象scene用来对规划场景进行操作
  moveit::planning_interface::PlanningSceneInterface scene;
  // 设置一个比例因子以选择性地降低最大关节速度限制，可取值为(0,1]
  arm.setMaxVelocityScalingFactor(0.8);
  gripper.setMaxVelocityScalingFactor(0.8);
  // 机械臂回到初始位置
  ROS_INFO("Moving to pose: Home");
  arm.setNamedTarget("Home");
  arm.move();
  // 设置桌面对象的形状和位姿
  moveit_msgs::CollisionObject table_object;
  table_object.header.frame_id = "base_link";
  table_object.id = "table";
  shape_msgs::SolidPrimitive table_primitive;
  table_primitive.type = table_primitive.BOX;
  table_primitive.dimensions.resize(3);
  table_primitive.dimensions[0] = 1.0;
  table_primitive.dimensions[1] = 1.2;
  table_primitive.dimensions[2] = 0.01;
  geometry_msgs::Pose table_pose;
  table_pose.position.x = 0;
  table_pose.position.y = 0;
  table_pose.position.z = -table_primitive.dimensions[2]/2.0;
  table_pose.orientation.w = 1.0;
  table_object.primitives.push_back(table_primitive);
  table_object.primitive_poses.push_back(table_pose);
  table_object.operation = table_object.ADD;

  // 设置长方体对象的形状和位姿
  moveit_msgs::CollisionObject box_object;
  box_object.header.frame_id = "base_link";
  box_object.id = "box";
  shape_msgs::SolidPrimitive box_primitive;
  box_primitive.type = box_primitive.BOX;
  box_primitive.dimensions.resize(3);
  box_primitive.dimensions[0] = 0.25;
  box_primitive.dimensions[1] = 0.3;
  box_primitive.dimensions[2] = 0.04;
  geometry_msgs::Pose box_pose;
  box_pose.position.x = 0.4;
  box_pose.position.y = 0;
  box_pose.position.z = box_primitive.dimensions[2]/2.0 + 0.37;
  box_pose.orientation.w = 1.0;
  box_object.primitives.push_back(box_primitive);
  box_object.primitive_poses.push_back(box_pose);
  box_object.operation = box_object.ADD;

  // 设置圆柱体对象的形状和位姿
  moveit_msgs::CollisionObject cylinder_object;
  cylinder_object.header.frame_id = "base_link";
  cylinder_object.id = "cylinder";
  shape_msgs::SolidPrimitive cylinder_primitive;
  cylinder_primitive.type = cylinder_primitive.CYLINDER;
  cylinder_primitive.dimensions.resize(2);
  cylinder_primitive.dimensions[0] = 0.12;
  cylinder_primitive.dimensions[1] = 0.015;
  geometry_msgs::Pose cylinder_pose;
  cylinder_pose.position.x = 0.4;
  cylinder_pose.position.y = 0.0;
  cylinder_pose.position.z = cylinder_primitive.dimensions[0]/2.0 + box_primitive.dimensions[2]/2.0 + 0.37;
  cylinder_pose.orientation.w = 1.0;
  cylinder_object.primitives.push_back(cylinder_primitive);
  cylinder_object.primitive_poses.push_back(cylinder_pose);
  cylinder_object.operation = cylinder_object.ADD;
  // 将桌面、长方体和圆柱体添加到collision_objects
  std::vector<moveit_msgs::CollisionObject> collision_objects;
  collision_objects.push_back(table_object);
  collision_objects.push_back(box_object);
  collision_objects.push_back(cylinder_object);
  // 将桌面、长方体和圆柱体添加到规划场景中
  scene.addCollisionObjects(collision_objects);
  scene.applyCollisionObjects(collision_objects);

  // 判断桌面是否已被添加到规划场景中
  if(waitForStateUpdate(table_object.id,scene,true)){
    ROS_INFO("The table has been successfully added.");
  }
  else{
    ROS_INFO("Failed to add the table.");
  }
  // 判断长方体是否已被添加到规划场景中
  if(waitForStateUpdate(box_object.id,scene,true)){
    ROS_INFO("The Box has been successfully added.");
  }
  else{
    ROS_INFO("Failed to add the Box.");
  }
  // 判断圆柱体是否已被添加到规划场景中
  if(waitForStateUpdate(cylinder_object.id,scene,true)){
    ROS_INFO("The Cylinder has been successfully added.");
  }
  else{
    ROS_INFO("Failed to add the Cylinder.");
  }

  ROS_INFO("Open gripper, move the arm and close gripper...");
  // 张开手爪
  std::vector<double> gripper_joint_positions = {0.65,0.65};
  gripper.setJointValueTarget(gripper_joint_positions);
  gripper.move();
  // 设置xarm规划组的目标位姿为圆柱体中心附近，让机械臂运动到该目标
  geometry_msgs::PoseStamped target_pose;
  target_pose.header.frame_id = "base_link";
  target_pose.header.stamp = ros::Time::now();
  target_pose.pose.position.x = 0.41;
  target_pose.pose.position.y = 0.0;
  target_pose.pose.position.z = cylinder_pose.position.z;
  target_pose.pose.orientation.w = 1;
  arm.setStartStateToCurrentState();
  arm.setPoseTarget(target_pose);
  arm.move();
  // 稍微闭合手爪
  gripper_joint_positions = {0.2,0.2};
  gripper.setJointValueTarget(gripper_joint_positions);
  gripper.move();
  ROS_INFO("Attach the Cylinder to the XBot-Arm robot ...");
  // 把cylinder附着到机械臂末端执行器上
  arm.attachObject(cylinder_object.id);
  // 判断圆柱体是否成功附着到了机械臂上
  if(waitForStateUpdate(cylinder_object.id,scene,false, true)){
    ROS_INFO("The Cylinder has been successfully attached.");
  }
  else{
    ROS_INFO("Failed to attach the Cylinder.");
  }
  ROS_INFO("Move the arm to a position 25 cm down ...");
  // 让机械臂运动到长方体下方的位置
  target_pose.pose.position.z -= 0.25;
  arm.setStartStateToCurrentState();
  arm.setPoseTarget(target_pose);
  arm.move();
  ROS_INFO("Detach the Cylinder from the XBot-Arm robot");
  // 将圆柱体与机械臂分离
  arm.detachObject(cylinder_object.id);
  sleep(2);
  ROS_INFO("Remove the objects from planning scene");
  // 删除规划场景里的物体对象
  std::vector<std::string> object_ids;
  object_ids.push_back(table_object.id);
  object_ids.push_back(box_object.id);
  object_ids.push_back(cylinder_object.id);
  scene.removeCollisionObjects(object_ids);
  sleep(1);
  // 机械臂回到初始位置
  ROS_INFO("Moving to pose: Home");
  arm.setNamedTarget("Home");
  arm.move();

  ros::shutdown();
  return 0;
}
